### ISSUE TITLE
Socket.SendFile() should fail with a consistent exception for connectionless sockets

### ISSUE THREAD

**Issue by antonfirsov** [dotnet/runtime#47472]
`Socket.SendFile()` is not supported on Windows, but the exceptions being thrown are not consistent. The first call throws `SocketException`, while the second attempt leads to `NotSupportedException`.

### Repro
The `twice=true` case fails:

```C#
[Theory]
[InlineData(false)]
[InlineData(true)]
public void UDP_SendFile(bool twice)
{
    string tempFile = Path.GetTempFileName();
    File.WriteAllBytes(tempFile, new byte[128]);

    using var client = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
    using var listener = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
    listener.BindToAnonymousPort(IPAddress.Loopback);

    client.Connect(listener.LocalEndPoint);

    Assert.Throws<SocketException>(() => client.SendFile(tempFile));
    if (twice)
    {
        Assert.Throws<SocketException>(() => client.SendFile(tempFile)); // Throws NotSupportedException
    }
}
```

## Edit (after discussion below)

On Linux the call does not fail. We should fail `NotSupportedException` for connectionless sockets on all OS-es in all cases.

**Comment by geoffkizer** [comment]
What do we do on Linux?

**Comment by geoffkizer** [comment]
related: #47287

**Comment by antonfirsov** [comment]
It does not throw. Maybe `sendfile` just works on Linux with UDP?

**Comment by gfoidl** [comment]
From the [manpage](https://man7.org/linux/man-pages/man2/sendfile.2.html):

> sendfile() copies data between one file descriptor and another.

This doesn't exclude UDP, so on linux it should work?

**Comment by wfurt** [comment]
of course with UDP, there is no way to now if what was sent is what was received. It would just blast content of the file without any feedback,

**Comment by antonfirsov** [comment]
I also think so (would be nice to double-check with a small app or test).

On Windows, it would be nice to throw consistent exceptions.

I wonder if we should explicitly invest efforts to document these platform differences better, or are these corner cases we don't care about much, since we don't expect users to meet them?

**Comment by geoffkizer** [comment]
Using SendFile with UDP doesn't really make much sense. The point of SendFile is to send large-ish files efficiently. Max size of a UDP packet is ~1500 bytes. It would be really weird to use SendFile with UDP.

Additionally, on Linux, the platform sendfile() API itself doesn't support pre and post buffers -- it just takes a fd. So we mimic the Windows behavior by sending the pre and post buffers manually using send(). But if you are using UDP, this probably isn't what you want -- the pre and post buffers will end up getting sent in different packets.

As such, I think we should just fail if you try to do SendFile on UDP for both Linux and Windows. The fact that this doesn't fail on Linux today is really an oversight.

Also, it seems fine to just always throw NotSupportedException here.

**Comment by antonfirsov** [comment]
Seems like we agree on this, I repurposed the issue based on the discussion outcome.

**Comment by ManickaP** [comment]
Triage: to keep is consistent we should throw on Linux, we should fix this for 6.0 until someone depends on not throwing.

**Comment by karelz** [comment]
Triage: Not a regression, UDP -- low usage. Fine to move to Future.
